lib/evse_security.cpp

```diff
@@ -356,37 +406,54 @@ EvseSecurity::get_installed_certificates(const std::vector<CertificateType>& cer
-            certificate_hash_data_chain.certificate_hash_data = certificate_hash_data;
-            certificate_hash_data_chain.certificate_type = CertificateType::V2GCertificateChain;
+            // Leaf V2G chain
+            X509CertificateBundle leaf_bundle(secc_key_pair.pair.value().certificate, EncodingFormat::PEM);
 
-            // TODO (ioan): as per V2GCertificateChain: OCPP 2.0.1 part 2 spec 2 (3.36):
-            // V2G certificate chain (excluding the V2GRootCertificate)
-            // Exclude V2G root when returning the V2GCertificateChain
+            // V2G chain
             const auto ca_bundle_path = this->ca_bundle_path_map.at(CaCertificateType::V2G);
             X509CertificateBundle ca_bundle(ca_bundle_path, EncodingFormat::PEM);
-            const auto certificates_of_bundle = ca_bundle.split();
-            std::vector<CertificateHashData> child_certificate_hash_data;
-
-            bool keep_searching = true;
-            while (keep_searching) {
-                keep_searching = false;
-                for (const auto& ca_cert : certificates_of_bundle) {
-                    if (X509_check_issued(ca_cert.get(), cert.get()) == X509_V_OK and
-                        ca_cert.get_issuer_name_hash() != cert.get_issuer_name_hash()) {
-                        CertificateHashData sub_ca_certificate_hash_data = ca_cert.get_certificate_hash_data();
-                        child_certificate_hash_data.push_back(sub_ca_certificate_hash_data);
-                        cert.reset(ca_cert.get());
-                        keep_searching = true;
-                        break;
+
+            // Merge the bundles
+            for (auto& certif : leaf_bundle.split()) {
+                ca_bundle.add_certificate_unique(std::move(certif));
+            }
+
+            // Create the certificate hierarchy
+            X509CertificateHierarchy& hierarchy = ca_bundle.get_certficate_hierarchy();
+            EVLOG_info << "Hierarchy:(V2GCertificateChain)\n" << hierarchy.to_debug_string();
+
+            // TODO (ioan): as per V2GCertificateChain: OCPP 2.0.1 part 2 spec 2 (3.36):
+            // V2G certificate chain (excluding the V2GRootCertificate)
+            // The first certificate is the 'leaf'
+            for (auto& root : hierarchy.get_hierarchy()) {
+                CertificateHashDataChain certificate_hash_data_chain;
+
+                certificate_hash_data_chain.certificate_type = CertificateType::V2GCertificateChain;
+
+                // Since the hierarchy starts with V2G and SubCa1/SubCa2 we have to add:
+                // the leaf as the first when returning
+                // * Leaf
+                // --- SubCa1
+                // --- SubCa2
+                std::vector<CertificateHashData> hierarchy_hash_data;
+
+                X509CertificateHierarchy::for_each_child(
+                    [&](const X509Node& child, int depth) { hierarchy_hash_data.push_back(child.hash); }, root);
```

Comment:
Is it guaranteed that the leaf will be the last certificate visited by this iteration? Is it possible that we somehow have multiple leaves (e.g. if the hierarchy is not linear, but has multiple branches)?

Reply:
Feedback from @AssemblyJohn : Additional leaves will not be included in this iteration due to `get_key_pair()` only accepting the latest valid leaf in the folder.